### ISSUE TITLE
Consider deprecating or changing the behaviour of CoroutineContext.isActive

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#3300]
According to the [doc](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/is-active.html), `isActive` has the following property:

>The coroutineContext.isActive expression is a shortcut for coroutineContext[Job]?.isActive == true. See [Job.isActive](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/is-active.html).

It means that, if the `Job` is not present, `isActive` always returns `false`.

We have multiple reports that such behaviour can be error-prone when used with non-`kotlinx.coroutines` entry points, such as Ktor and `suspend fun main`, because it is inconsistent with the overall contract:

>(Job) has not been completed and was not cancelled yet

`CoroutineContext.isActive` predates both `CoroutineScope` (which should always have a `Job` in it, if it's not `GlobalScope`)  and `job` extension, so it may be the case that it can be safely deprecated.

Basically, we have three options:

* Do nothing, left things as is. It doesn't solve the original issue, but also doesn't introduce any potentially breaking changes
* Deprecate `CoroutineContext.isActive`. Such change has multiple potential downsides
    * Its only possible replacement is `this.job.isActive`, but this replacement is not equivalent to the original method -- `.job` throws an exception for contexts without a `Job`. An absence of replacement can be too disturbing as [a lot of code](https://grep.app/search?q=context.isActive&filter[lang][0]=Kotlin) rely on a perfectly fine `ctxWithJob.isActive`
    * Code that relies on `.job.isActive` no longer can be called from such entry points safely
* Change the default behaviour -- return `true`. It also "fixes" such patterns as `GlobalScope.isActive` but basically is a breaking change

**Comment by joffrey-bion** [comment]
> Change the default behaviour -- return true. It also "fixes" such patterns as GlobalScope.isActive but basically is a breaking change

`GlobalScope.isActive` is not the same property, and already defaults to `true` when there is no job, so this one is not broken:
https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/CoroutineScope.kt#L137

This is why I believe it's often more correct to use `CoroutineScope.isActive` rather than `CoroutineContext.isActive`.

> Its only possible replacement is `this.job.isActive`

Why, though? Isn't `this[Job]?.isActive == true` another perfectly valid replacement that doesn't change behaviour? Or is this what you mean by "absence of replacement" - just inlining the property?

I would be reluctant in suggesting the `.job.isActive` replacement. Very often the authors of those pieces of code just didn't think of the jobless case, and here we would push them towards making their code fail even harder in this case (possibly unknowingly).

> An absence of replacement can be too disturbing as [a lot of code](https://grep.app/search?q=context.isActive&filter%5Blang%5D%5B0%5D=Kotlin) rely on a perfectly fine `ctxWithJob.isActive`

By deprecating this property instead of changing the default, users could "fix" their code by wrapping the body of their function with `coroutineScope { ... }` and accessing the `CoroutineScope.isActive` property from that scope (which would now have a job). Or just use `yield()` if they want to cancel a non-suspending loop. Or just remove that condition altogether because they didn't need it in the first place (calling other suspend functions with prompt cancellation guarantees already).

It's a matter of whether we want `CoroutineContext.isActive` to make sense without jobs I guess.
---
TL;DR: I like both the deprecation option and the default change option, but if we deprecate I find the replacement `.job.isActive` harmful
